# Hola Cece!

Mi nombre es David Bautista, soy code reviewer de Tripleten y hoy tengo el gusto de revisar tu proyecto.

Cuando vea un error laa primera vez, lo señalaré. Deberás encontrarlo y arreglarlo. La intención es que te prepares para un espacio real de trabajo. En un trabajo, el líder de tu equipo hará lo mismo. Si no puedes solucionar el error, te daré más información en la próxima ocasión.

Encontrarás mis comentarios más abajo - por favor, no los muevas, no los modifiques ni los borres.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>


<div class="alert alert-block alert-danger">
    
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>


Puedes responderme de esta forma: 

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div>

¡Empecemos!

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Objetivos" data-toc-modified-id="Objetivos-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Objetivos</a></span></li><li><span><a href="#Inicialización" data-toc-modified-id="Inicialización-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Inicialización</a></span></li><li><span><a href="#Conexión-a-la-base-de-datos" data-toc-modified-id="Conexión-a-la-base-de-datos-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Conexión a la base de datos</a></span></li><li><span><a href="#Exploración-de-tablas" data-toc-modified-id="Exploración-de-tablas-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Exploración de tablas</a></span></li><li><span><a href="#Ejercicios" data-toc-modified-id="Ejercicios-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Ejercicios</a></span><ul class="toc-item"><li><span><a href="#Número-de-libros-publicados-después-del-1-de-enero-de-2000" data-toc-modified-id="Número-de-libros-publicados-después-del-1-de-enero-de-2000-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Número de libros publicados después del 1 de enero de 2000</a></span></li><li><span><a href="#Número-de-reseñas-de-usuarios-y-la-calificación-promedio-para-cada-libro" data-toc-modified-id="Número-de-reseñas-de-usuarios-y-la-calificación-promedio-para-cada-libro-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Número de reseñas de usuarios y la calificación promedio para cada libro</a></span></li><li><span><a href="#Editorial-que-ha-publicado-el-mayor-número-de-libros-con-más-de-50-páginas" data-toc-modified-id="Editorial-que-ha-publicado-el-mayor-número-de-libros-con-más-de-50-páginas-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Editorial que ha publicado el mayor número de libros con más de 50 páginas</a></span></li><li><span><a href="#autor-que-tiene-la-más-alta-calificación-promedio-del-libro:-mira-solo-los-libros-con-al-menos-50-calificaciones." data-toc-modified-id="autor-que-tiene-la-más-alta-calificación-promedio-del-libro:-mira-solo-los-libros-con-al-menos-50-calificaciones.-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.</a></span></li><li><span><a href="#Número-promedio-de-reseñas-de-texto-entre-los-usuarios-que-calificaron-más-de-50-libros." data-toc-modified-id="Número-promedio-de-reseñas-de-texto-entre-los-usuarios-que-calificaron-más-de-50-libros.-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.</a></span></li></ul></li></ul></div>

# Descripción del Proyecto


El coronavirus tomó al mundo entero por sorpresa, cambiando la rutina diaria de todos y todas. Los habitantes de las ciudades ya no pasaban su tiempo libre fuera, yendo a cafés y centros comerciales; sino que más gente se quedaba en casa, leyendo libros. Eso atrajo la atención de las startups (empresas emergentes) que se apresuraron a desarrollar nuevas aplicaciones para los amantes de los libros.

Te han dado una base de datos de uno de los servicios que compiten en este mercado. Contiene datos sobre libros, editoriales, autores y calificaciones de clientes y reseñas de libros. Esta información se utilizará para generar una propuesta de valor para un nuevo producto.

## Objetivos 


1. Identificar el número de libros publicados después del 1 de enero de 2000.
2. Determinar el número de reseñas de usuarios y la calificación promedio para cada libro.
3. Identificar la editorial que ha publicado el mayor número de libros con más de 50 páginas.
4. Identificar al autor con la calificación promedio más alta para los libros que tienen al menos 50 calificaciones.
5. Encontrar el número promedio de reseñas de texto entre los usuarios que han calificado más de 50 libros.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, Cece, muy buen trabajo con toda la sección de introducción del proyecto. Usas una tabla de contenido, realizas una descripción del proyecto, y mencionas objetivos del estudio, realmente buen trabajo. </div>

## Inicialización

A continuación cargaré las librerias necesarias para el desarrollo del proyecto.

In [1]:
#!python -m pip install --upgrade 'sqlalchemy<2.0'

In [2]:
#pip install psycopg2

In [3]:
import pandas as pd
from sqlalchemy import create_engine
import psycopg2

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo importando las librerías necesarias para el desarrollo del proyecto. </div>

## Conexión a la base de datos

Obtener acceso al conjunto de datos, revisar la estructura del conjunto de datos, identificar variables relevantes y entender la calidad de los datos.

In [4]:

db_config = {'user': 'practicum_student', # nombre de usuario
'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # puerto de conexión
'db': 'data-analyst-final-project-db'} # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],

db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo configurando la conexión al servidor ``SQL``. </div>

## Exploración de tablas

In [5]:
#books
query= "SELECT * FROM books LIMIT 5"
pd.read_sql(query, engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
#authors
query= "SELECT * FROM authors LIMIT 5"
pd.read_sql(query, engine)


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
#publishers
query= "SELECT * FROM publishers LIMIT 5"
pd.read_sql(query, engine)


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [37]:
#ratings
query= "SELECT * FROM ratings LIMIT 5"
pd.read_sql(query, engine)


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [9]:
#reviews
query= "SELECT * FROM reviews LIMIT 5"
pd.read_sql(query, engine)


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, Cece, buen trabajo con esta primera vista de las tablas. </div>

## Ejercicios

### Número de libros publicados después del 1 de enero de 2000

In [10]:
query = """
        SELECT COUNT(*) AS num_libros
        FROM books
        WHERE publication_date > '2000-01-01';
        """
pd.read_sql(query, engine)


num_libros
0         819

Hay 819 libros publicados después del 1 de enero de 2000

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo identificando la cantidad de libros publicados después del 1 de enero del 2000.</div>

### Número de reseñas de usuarios y la calificación promedio para cada libro

In [11]:
query = """
SELECT
    b.book_id,
    COALESCE(COUNT(r.rating_id), 0) AS num_reseñas,
    COALESCE(AVG(r.rating), 0) AS calificacion_promedio
FROM
    books b
LEFT JOIN
    ratings r ON b.book_id = r.book_id
GROUP BY
    b.book_id;
        """
pd.read_sql(query, engine)


book_id  num_reseñas  calificacion_promedio
0        652            2               4.500000
1        273            2               4.500000
2         51           12               4.250000
3        951            2               4.000000
4        839            7               4.285714
..       ...          ...                    ...
995       64           13               4.230769
996       55            2               5.000000
997      148            7               3.428571
998      790            2               3.500000
999      828            2               3.000000

[1000 rows x 3 columns]

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo creando la tabla. </div>

### Editorial que ha publicado el mayor número de libros con más de 50 páginas 

In [12]:
query = """
SELECT
    p.publisher_id,
    p.publisher,
    COUNT(CASE WHEN b.num_pages > 50 THEN b.book_id END) AS num_libros
FROM
    books b
JOIN
    publishers p ON b.publisher_id = p.publisher_id
GROUP BY
    p.publisher_id, p.publisher
ORDER BY
    num_libros DESC
LIMIT 1;
"""
pd.read_sql(query, engine)


publisher_id      publisher  num_libros
0           212  Penguin Books          42

La editorial que ha publicado el mayor número de libros con más de 50 páginas es Penguin Books	

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, ``Penguin Books`` es la editorial que ha publivado el mayor número de libros con más de 50 páginas.</div>

### autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.

In [26]:
query = """
SELECT
    a.author_id,
    a.author,
    AVG(r.rating) AS calificacion_promedio
FROM
    authors a
JOIN
    books b ON a.author_id = b.author_id
JOIN
    ratings r ON b.book_id = r.book_id
GROUP BY
    a.author_id, a.author
HAVING
    COUNT(r.rating_id) >= 50
ORDER BY
    calificacion_promedio DESC
LIMIT 1;
"""
pd.read_sql(query, engine)



author_id          author  calificacion_promedio
0        130  Diana Gabaldon                    4.3

El autor que tiene la más alta calificación promedio del libro es Diana Gabaldon con 4.3	

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo Cece.</div>

### Número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [14]:
query = """
SELECT
    AVG(n_reviews_text) AS promedio_reseñas_texto
FROM (
    SELECT
        username,
        COUNT(review_id) AS n_reviews_text
    FROM
        reviews
    WHERE
        username IN (
            SELECT
                username
            FROM
                ratings
            GROUP BY
                username
            HAVING
                COUNT(rating_id) > 50
        )
    GROUP BY
        username
) AS subquery;
"""
pd.read_sql(query, engine)

promedio_reseñas_texto
0               24.333333

En promedio cada usuario ha escrito alrededor de 24 reseñas.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo realizando el calculo.</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
# Comentario General
    
Hola, Cece, te felicito por el desarrollo del proyecto. Realizaste un muy buen trabajo.  </div>